### PR TITLE
test: Convert amounts from float to decimal

### LINKED ISSUE #20011: test: Convert amounts in testmempoolaccept test from float to decimal
Float should not be used for currency amounts (not even in tests). Thus `float` should be converted to use `decimal`. See also https://github.com/bitcoin/bitcoin/pull/19940#pullrequestreview-491979459

#### Useful skills:

Python 3

#### Want to work on this issue?

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by ghost**
> decimal is preferred in accounting applications

https://docs.python.org/3.8/library/decimal.html

Decimal type saves an exact value so better than using float.

~~3 variables declared with type as 'Decimal' in [test/functional/mempool_accept.py](https://github.com/bitcoin/bitcoin/blob/master/test/functional/mempool_accept.py): fee, fee_expected, output_amount~~

~~Not required to convert to string anymore for using the above variables as decimal~~

+ fee, fee_expected, output_amount
~~+ 8 decimal places~~
+ Using value of coin['amount'] as decimal and removed 'int'
+ Removed unnecessary parentheses
+ Remove str() and use quotes

Fixes https://github.com/bitcoin/bitcoin/issues/20011

### CODE REVIEW

File: test/functional/mempool_accept.py [hunk 1]
```diff
@@ -83,7 +83,7 @@ def run_test(self):
         )
 
         self.log.info('A transaction not in the mempool')
-        fee = 0.00000700
+        fee: Decimal = 0.00000700
```
Comment (luke-jr) [inline test/functional/mempool_accept.py:86]: Where is this syntax defined? It seems very ugly/weird for Python, and in my Python 3.7.8 console, it also doesn't result in a Decimal type...

File: test/functional/mempool_accept.py [hunk 2]
```diff
@@ -92,22 +92,22 @@ def run_test(self):
         tx.deserialize(BytesIO(hex_str_to_bytes(raw_tx_0)))
         txid_0 = tx.rehash()
         self.check_mempool_result(
-            result_expected=[{'txid': txid_0, 'allowed': True, 'vsize': tx.get_vsize(), 'fees': {'base': Decimal(str(fee))}}],
+            result_expected=[{'txid': txid_0, 'allowed': True, 'vsize': tx.get_vsize(), 'fees': {'base': (fee)}}],
```
Comment (luke-jr) [inline test/functional/mempool_accept.py:95]: Prefer to drop the unnecessary parenthesis here.

File: test/functional/mempool_accept.py (hunk 1 above)
Comment (sipa) [reply test/functional/mempool_accept.py:86]: It's a type annotation. It has no operational semantics (so execution will be identical to not having them at all). All it does is fail type checking if that's enabled.

I assume OP is incorrectly assuming that it actually determines the runtime type.

Comment (prayank23) [reply test/functional/mempool_accept.py:86]: I found it here: https://docs.python.org/3/library/typing.html

Comment (laanwj) [reply test/functional/mempool_accept.py:86]: I don't think this is correct. Python never interprets inline numbers as Decimal, and the type annotation doesn't change that. I think what you want is:
```python
fee: Decimal = Decimal('0.00000700')
```

Comment (prayank23) [reply test/functional/mempool_accept.py:86]: As sipa mentioned, I assumed python has introduced something like few other languages which I missed. I should have read the details and tested few things.

@laanwj your suggestion is correct however I am not sure if we should use the type annotations if they don't change anything. Planning to make a new commit with changes like 'test4' mentioned in the below pic after trying few things on my system and doing more research:

image

**Comment by prayank23** [comment]
Made the changes in https://github.com/bitcoin/bitcoin/pull/20039/commits/ba15b848d9d9bb05f5e1b655ab417ac9654df32e

Was even thinking about adding `getcontext().prec = 8` so that all operations in this file on decimals eventually result in 8 decimal places but not sure if it's the best approach to be used in bitcoin core so didn't add it.

File: test/functional/mempool_accept.py [hunk 3]
```diff
@@ -190,7 +190,7 @@ def run_test(self):
         tx.deserialize(BytesIO(hex_str_to_bytes(raw_tx_reference)))
         # Reference tx should be valid on itself
         self.check_mempool_result(
-            result_expected=[{'txid': tx.rehash(), 'allowed': True, 'vsize': tx.get_vsize(), 'fees': { 'base': Decimal(str(0.1 - 0.05))}}],
+            result_expected=[{'txid': tx.rehash(), 'allowed': True, 'vsize': tx.get_vsize(), 'fees': { 'base': Decimal(str(0.10000000 - 0.05000000))}}],
```
Comment (guggero) [inline test/functional/mempool_accept.py:193]: These are the values from L170 and L188. I think we should extract those to a proper variable that is also of type `Decimal`. Or at the very least this line should read `... { 'base': Decimal('0.10000000') - Decimal('0.05000000')}` IMO.

**Review by guggero — COMMENTED** [review]
Concept ACK 3f56b32c.

**Comment by decryp2kanon** [comment]
Concept ACK 3f56b32

File: test/functional/mempool_accept.py (hunk 3 above)
Comment (prayank23) [reply test/functional/mempool_accept.py]: Made the changes to remove `str()` in 7be315edd7bb5dbc4c27052b3f38d1498b51ce8b

**Comment by kristapsk** [comment]
Concept ACK, but should squash the commits, I think.

File: test/functional/mempool_accept.py [hunk 4]
```diff
@@ -83,31 +83,31 @@ def run_test(self):
         )
 
         self.log.info('A transaction not in the mempool')
-        fee = 0.00000700
+        fee = Decimal('0.00000700')
         raw_tx_0 = node.signrawtransactionwithwallet(node.createrawtransaction(
             inputs=[{"txid": txid_in_block, "vout": 0, "sequence": BIP125_SEQUENCE_NUMBER}],  # RBF is used later
-            outputs=[{node.getnewaddress(): 0.3 - fee}],
+            outputs=[{node.getnewaddress(): Decimal('0.3000000') - fee}],
```
Comment (guggero) [inline test/functional/mempool_accept.py:89]: This is missing a zero, only 7 decimal places are shown.

Comment (prayank23) [reply test/functional/mempool_accept.py]: Fixed. Now has 8 decimal places.

**Review by guggero — COMMENTED** [review]
ACK 3afdc9bd - assuming commits are being squashed on merge.

File: test/functional/mempool_accept.py [hunk 5]
```diff
@@ -83,31 +83,31 @@ def run_test(self):
         )
 
         self.log.info('A transaction not in the mempool')
-        fee = 0.00000700
+        fee = Decimal('0.00000700')
         raw_tx_0 = node.signrawtransactionwithwallet(node.createrawtransaction(
             inputs=[{"txid": txid_in_block, "vout": 0, "sequence": BIP125_SEQUENCE_NUMBER}],  # RBF is used later
-            outputs=[{node.getnewaddress(): 0.3 - fee}],
+            outputs=[{node.getnewaddress(): Decimal('0.30000000') - fee}],
```
Comment (MarcoFalke) [inline test/functional/mempool_accept.py:89]: Trailing zeros are ignored, so you can leave them out

**Comment by MarcoFalke** [comment]
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

File: test/functional/mempool_accept.py (hunk 5 above)
Comment (prayank23) [reply test/functional/mempool_accept.py:89]: @MarcoFalke so should I remove trailing zeros for all or keep the things as they are right now after last commit in this PR?

Comment (MarcoFalke) [reply test/functional/mempool_accept.py:89]: I'd prefer to remove them, unless there is a reason to add them

**Comment by guggero** [comment]
ACK 5aadd4be1883386a04bef6a04e9a1142601ef7a7

**Pull request merged**